### PR TITLE
{DO NOT MERGE} [CDEC-512] Create JSON round-trip and golden tests for GenesisData, GenesisProtocolConstants, and GenesisConfiguration

### PR DESCRIPTION
**Pull request by intricate**
## Description

We have to create and expand upon existing round-trip and golden tests for the aforementioned data types in order to ensure that the upcoming implementation of [CO-354](https://iohk.myjetbrains.com/youtrack/issue/CO-354) doesn't break configuration backwards compatibility.

## Linked issue

[https://iohk.myjetbrains.com/youtrack/issue/CDEC-512](https://iohk.myjetbrains.com/youtrack/issue/CDEC-512)
## Type of change

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [x] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [ ] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.

## Testing checklist

- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: util/test/Test/Pos/Util/Golden.hs [hunk 1]
```diff
@@ -38,6 +43,35 @@ goldenTestJSON x path = withFrozenCallStack $ do
             Left err -> failWith Nothing $ "could not decode: " <> show err
             Right x' -> x === x'
 
+-- | Only check that the datatype equals the decoding of the file
```
Comment (Jimbo4350) [inline util/test/Test/Pos/Util/Golden.hs]: Shouldn't we also be checking the encoding?
https://github.com/input-output-hk/cardano-sl/pull/3415

File: util/test/Test/Pos/Util/Golden.hs [hunk 2]
```diff
@@ -38,6 +43,35 @@ goldenTestJSON x path = withFrozenCallStack $ do
...
+goldenTestCanonicalJSONDec
+    :: ( Eq a
+       , Canonical.FromJSON (Either SchemaError) a
+       , HasCallStack
+       , Show a
+       )
+    => a
+    -> FilePath
+    -> Property
+goldenTestCanonicalJSONDec x path = withFrozenCallStack $ do
```
Comment (Jimbo4350) [inline util/test/Test/Pos/Util/Golden.hs]: Same question as above.

File: util/test/Test/Pos/Util/Golden.hs (hunk 1 above)
Comment (mhuesch) [reply util/test/Test/Pos/Util/Golden.hs]: The reason for these functions is to ensure backwards compatibility with existing serialized datatstructures, with default values for missing fields. We want to demonstrate that when the PR is merged, the tests for “deserialize this bytestring and check that the value is equal to a target value” still pass (once the given datatype is modified, in the source code, to have the default value in the newly added record field)

File: util/test/Test/Pos/Util/Golden.hs (hunk 2 above)
Comment (mhuesch) [reply util/test/Test/Pos/Util/Golden.hs]: See above comment. These should be used sparingly, only to ensure backwards compatibility.

**Comment by mhuesch** [comment]
[Closing this, as we chose a different approach for CO-354](https://iohk.myjetbrains.com/youtrack/issue/CDEC-512#comment=93-27918).

**Pull request closed without merging**